test/common/network_policy_access_ns_to_svc.go

```diff
@@ -0,0 +1,278 @@
+	}
+
+	// creating a project
+	_, err = createProject(ctx, openshiftAPIURL, token)
+	if err != nil {
+		t.Fatalf("error occured while creating a project: %v", err)
+	}
+
+	// creating service
+	serviceCR, err := createService(ctx, openshiftAPIURL, token)
+	if err != nil {
+		t.Fatalf("error occured while creating a service: %v", err)
+	}
+
+	// creating pod
+	podCR, err := createPodWithAnEndpoint(ctx, openshiftAPIURL, token)
+	if err != nil {
+		t.Fatalf("error occured while creating a pod: %v", err)
+	}
+
+	podReady, err := checkPodStatus(ctx, podCR)
+	if err != nil {
+		t.Fatalf("error checking pod status: %v", err)
+	}
+
+	if podReady == false {
+		t.Fatalf("pod %s failed to become ready", podCR.GetName())
+	}
+
+	tsApicastPod, err := get3ScaleAPIcastPod(ctx)
+	if err != nil {
+		t.Fatalf("error getting 3scale apicast pod: %v", err)
+	}
+
+	curlCommand := fmt.Sprintf("curl %s.%s.svc.cluster.local", serviceCR.GetName(), serviceCR.GetNamespace())
+	apicastContainerName := "apicast-production"
+	outputCurlCommand, err := execToPod(curlCommand,
+		tsApicastPod.GetName(),
+		tsApicastPod.GetNamespace(),
+		apicastContainerName,
+		ctx)
+	if err != nil {
+		t.Fatalf("error occured while executing curl command in pod: %v", err)
+	}
+
+	if strings.TrimSpace(outputCurlCommand) != podEndpoitResponse {
+		t.Fatalf("Failed to validate response Pod apicast received %s, instead of %s", outputCurlCommand, podEndpoitResponse)
+	}
+}
+
+func getOauthToken(ctx *TestingContext, apiUrl string, username string, password string) (string, error) {
+
+	// get oauth route
+	oauthRoute := &routev1.Route{}
+	key := types.NamespacedName{
+		Name:      resources.OpenshiftOAuthRouteName,
+		Namespace: resources.OpenshiftAuthenticationNamespace,
+	}
+
+	if err := ctx.Client.Get(goctx.TODO(), key, oauthRoute); err != nil {
+		return "", fmt.Errorf("error getting Openshift Oauth Route: %s", err)
+	}
+
+	// get dedicated admin token
+	dedicatedAdminToken, err := resources.DoAuthOpenshiftUser(
+		oauthRoute.Spec.Host,
+		apiUrl,
+		resources.DefaultIDP,
+		username,
+		password,
+	)
+	if err != nil {
+		return "", fmt.Errorf("error occured trying to get token : %v", err)
+	}
+
+	return dedicatedAdminToken, nil
+}
+
+// creating project as dedicated admin
+func createProject(ctx *TestingContext, openshiftAPIURL string, token string) (*projectv1.ProjectRequest, error) {
+	projectCR := &projectv1.ProjectRequest{
+		ObjectMeta: v1.ObjectMeta{
+			Name: projectName,
+		},
+	}
+
+	// check if project exist
+	err := ctx.Client.Get(goctx.TODO(), k8sclient.ObjectKey{Name: projectName}, &projectv1.Project{})
+	if err != nil && !k8serr.IsNotFound(err) {
+		return nil, fmt.Errorf("error occured while retrieving a project: %v", err)
+	} else if k8serr.IsNotFound(err) {
+		if err := resources.DoOpenshiftCreateProject(openshiftAPIURL, token, projectCR); err != nil {
+			return nil, fmt.Errorf("error occured while making request to create a project: %v", err)
+		}
+	}
```

Comment:
I wonder would this be better write as below. The `k8serr.IsNotFound` the one you care most about and I think you should check for that first.
```
	if k8serr.IsNotFound(err) {
		if err := resources.DoOpenshiftCreateProject(openshiftAPIURL, token, projectCR); err != nil {
			return nil, fmt.Errorf("error occured while making request to create a project: %v", err)
		}
	} else if err != nil {
		return nil, fmt.Errorf("error occured while retrieving a project: %v", err)
	}
```

Reply:
Makes sense but after having a better look at the code I think we can get rid of the `else` statement as it'll only progress if the resources was not found, wdyt?

```
	err := ctx.Client.Get(goctx.TODO(), key, serviceCR)
	if err != nil && !k8serr.IsNotFound(err) {
		return nil, fmt.Errorf("error occured while retrieving service %s : %v", serviceCR.GetName(), err)
	}

	err = resources.DoOpenshiftCreateServiceInANamespace(openshiftAPIURL, token, projectName, serviceCR)
	if err != nil {
		return nil, fmt.Errorf("error occured while making request to create service %s : %v", serviceCR.GetName(), err)
	}
```

Reply:
I was thinking here after your suggestion and I am not sure if it really matter much as `!k8serr.IsNotFound(err)` will be only checked if `err` is not nil, but for the sake of readability I always prefer to handle the errors after an operation.